File: dump-cn/489.robot-room-cleaner.cpp
class Solution {
public:
	vector<int> xV{ -1,0,1,0 };
	vector<int> yV{ 0,1,0,-1 };
	void dfsRobot(Robot& robot, unordered_set<string>& visited, int x, int y, int dir) {
		string key = to_string(x) + ',' + to_string(y);
		if (visited.count(key))
			return;
		visited.insert(key);
		robot.clean();
		for (int i = 0; i < 4; ++i) {
			// 先至少转一个方向
			int newDir = (dir + i + 1) % 4;
			int x1 = x + xV[newDir];
			int y1 = y + yV[newDir];
			// 转向后如果面对障碍物 进入下一个方向
			robot.turnRight();
			if (!robot.move())
				continue;
			// 下层遍历
			dfsRobot(robot, visited, x1, y1, newDir);
			// 遍历完返回 转180度后走一步即可
			robot.turnRight();
			robot.turnRight();
			robot.move();
			// 返回后记得要把朝向转过来 再转180度
			robot.turnRight();
			robot.turnRight();
		}
	}

	void cleanRoom(Robot& robot) {
		unordered_set<string> visited;
		dfsRobot(robot, visited, 0, 0, 0);
	}
};

File: dump-cn/489.robot-room-cleaner.go
func cleanRoom(robot *Robot) {
    // serialize pair of row and column number as key
    visited := make(map[string]bool)
    // generate clockwise directions, order matters
    // order should match with robot direction
    rowDirs := []int{-1,0,1,0}
    colDirs := []int{0,1,0,-1}
    var backtrack func(row, column, direction int)
    backtrack = func(row, column, direction int){
        // no end condition, we DFS until all recursion is finished
        k := fmt.Sprintf("%d,%d",row,column)
        visited[k] = true
        robot.Clean()
        for i := 0; i < 4; i++{
            // use module otherwise newD could cause invalid index of Dirs
            newD := (i+direction) % 4
            newRow := row + rowDirs[newD]
            newCol := column + colDirs[newD]
            newK := fmt.Sprintf("%d,%d",newRow,newCol)
            if visited[newK] || !robot.Move() {
                robot.TurnRight()
                continue
            }
            // select current choice
            backtrack(newRow,newCol,newD)
            // once finished, revert
            // first back track
            robot.TurnRight()
            robot.TurnRight()
            robot.Move()
            // then return to previous direction
            // robot.TurnLeft();
            // robot.TurnLeft();
            robot.TurnRight()
            robot.TurnRight()
            
            robot.TurnRight()
        }
    }
    backtrack(0,0,0)
}